### PR TITLE
Dict and Http functions

### PR DESCRIPTION
**Pull request by mariajdab**
## What is the problem/goal being addressed?
Added functions to `LibDict`,  `LibtHttp` and the test cases in `dict.test` and `http.test`
## What is the solution to this problem?
Added functions to `LibDict`,  `LisbHttp` and the test cases in `dict.test` and `http.test`

## How are you sure this works/how was this tested?
Test cases passed

## Has this information been included in the comments?
Yes

### CODE REVIEW

File: fsharp-backend/src/LibExecution/StdLib/LibDict.fs [hunk 1]
```diff
@@ -88,154 +90,57 @@ let fns : List<BuiltInFn> =
...
+      description =
+        "Returns a new dict with `entries`. Each value in `entries` must be a `[key, value]` list, where `key` is a `String`.
+        If `entries` contains duplicate `key`s, the last entry with that key will be used in the resulting dictionary (use `Dict::fromList` if you want to enforce unique keys).
+        This function is the opposite of `Dict::toList`."
+      fn =
+        (function
+        | state, [DList l] ->
+          let f acc e =
+            match e with
+              | DList [DStr k; value] when Map.containsKey k acc -> Map.remove k acc |> Map.add k value
```
Comment (pbiggar) [inline fsharp-backend/src/LibExecution/StdLib/LibDict.fs:105]: This condition is not needed, as `add` will overwrite an existing key

File: fsharp-backend/src/LibExecution/StdLib/LibDict.fs [hunk 2]
```diff
@@ -88,154 +90,57 @@ let fns : List<BuiltInFn> =
...
+              | DList [DStr k; value] -> Map.add k value acc
+              | DList [k; value] -> Errors.throw(Errors.argumentWasnt "String`s" "key" k)
+              | _ -> Errors.throw "All list items must be `[key, value]`"
```
Comment (pbiggar) [inline fsharp-backend/src/LibExecution/StdLib/LibDict.fs:108]: There's an error condition in the old code, where if there is a DIncomplete, DErorrRail or DError found, it is returned. You can replicate that here using `Errors.foundFakeDval`. (Please add a test for it too, if possible)

File: fsharp-backend/src/LibExecution/StdLib/LibDict.fs [hunk 3]
```diff
@@ -88,154 +90,57 @@ let fns : List<BuiltInFn> =
...
+        If `entries` contains duplicate `key`s, the last entry with that key will be used in the resulting dictionary (use `Dict::fromList` if you want to enforce unique keys).
+        This function is the opposite of `Dict::toList`."
+      fn =
+        (function
+        | state, [DList l] ->
+          let f acc e =
+            match e with
+              | DList [DStr k; value] when Map.containsKey k acc -> Map.remove k acc |> Map.add k value
+              | DList [DStr k; value] -> Map.add k value acc
+              | DList [k; value] -> Errors.throw(Errors.argumentWasnt "String`s" "key" k)
```
Comment (pbiggar) [inline fsharp-backend/src/LibExecution/StdLib/LibDict.fs:107]:
```suggestion
              | DList [k; value] -> Errors.throw(Errors.argumentWasnt "a string" "key" k)
```

File: fsharp-backend/src/LibExecution/StdLib/LibDict.fs [hunk 4]
```diff
@@ -88,154 +90,57 @@ let fns : List<BuiltInFn> =
...
+         Otherwise, returns `Nothing` (use `Dict::fromListOverwritingDuplicates` if you want to overwrite duplicate keys)."
+      fn =
+        (function
+        | state, [ DList l ] ->
+
+          let f acc e =
+            match acc, e with
+              | Some acc, DList [DStr k; value] when Map.containsKey k acc -> None
+              | Some acc, DList [DStr k; value] -> Some(Map.add k value acc)
+              | Some acc, DList [k; value] -> Errors.throw(Errors.argumentWasnt "String`s" "key" k)
```
Comment (pbiggar) [inline fsharp-backend/src/LibExecution/StdLib/LibDict.fs:131]:
```suggestion
              | Some acc, DList [k; value] -> Errors.throw(Errors.argumentWasnt "a string" "key" k)
```

File: fsharp-backend/tests/testfiles/dict.tests [hunk 5]
```diff
@@ -8,18 +8,19 @@ Dict.empty_v0 = {}
...
-// Dict.fromListOverwritingDuplicates_v0 [["a";1];["b";2];["a";3]] = { b = 2 ; a = 3 }
-// Dict.fromListOverwritingDuplicates_v0 [["a";1];["b";2];["c";3]] = { c = 3 ; b = 2 ; a= 1 }
-// Dict.fromListOverwritingDuplicates_v0 [] = {}
-// Dict.fromListOverwritingDuplicates_v0 [["a";1];["b";2];["c";3;3]] = Test.typeError_v0 "Todo: It has length 3 but must have length 2"
-// Dict.fromListOverwritingDuplicates_v0 [[1;1]] = Test.typeError_v0 "Todo: Keys must be Strings"
+Dict.fromListOverwritingDuplicates_v0 [["a";1];["b";2];["a";3]] = { b = 2 ; a = 3 }
+Dict.fromListOverwritingDuplicates_v0 [["a";1];["b";2];["c";3]] = { c = 3 ; b = 2 ; a= 1 }
+Dict.fromListOverwritingDuplicates_v0 [] = {}
+//Dict.fromListOverwritingDuplicates_v0 [["a";1];["b";2];["c";3;3]] = Test.typeError_v0 "Todo: It has length 3 but must have length 2"
+Dict.fromListOverwritingDuplicates_v0 [[1;1]] = Test.typeError_v0 "Expected the argument `key` to be String`s, but it was `1`" // FSHARPONLY
```
Comment (pbiggar) [inline fsharp-backend/tests/testfiles/dict.tests:15]: Can you add an equivalent OCAMLONLY test?

File: fsharp-backend/tests/testfiles/dict.tests [hunk 6]
```diff
@@ -8,18 +8,19 @@ Dict.empty_v0 = {}
...
 // Dict.fromListOverwritingDuplicates_v0 [["a";1];2;["c";3]] = Test.typeError_v0 "Todo: It is of type Int instead of List"
-// Dict.fromList_v0 [["a";1];["b";2];["a";3]] = Nothing // duplicate key
-// Dict.fromList_v0 [["a";1];["b";2];["c";3]] = Just { c = 3; b = 2; a = 1 }
+Dict.fromList_v0 [["a";1];["b";2];["a";3]] = Nothing // duplicate key
+Dict.fromList_v0 [["a";1];["b";2];["c";3]] = Just { c = 3; b = 2; a = 1 }
+Dict.fromList_v0 [ ["Content-Length";0]; ["Server"; "dark"] ] = Just {``Content-Length`` = 0; Server = "dark"}
 // Dict.fromList_v0 [["a";1];["b";2];["c";3;3]] = Test.typeError_v0 "Todo: It has length 3 but must have length 2"
-// Dict.fromList_v0 [[1;1]] = Test.typeError_v0 "Todo: Keys must be Strings"
+Dict.fromList_v0 [[1;1]] = Test.typeError_v0 "Expected the argument `key` to be String`s, but it was `1`" // FSHARPONLY
```
Comment (pbiggar) [inline fsharp-backend/tests/testfiles/dict.tests:21]: Can you add an equivalent OCAMLONLY test?

**Review by pbiggar — CHANGES_REQUESTED** [review]
Looks great! Should be good to go with some minor changes.

**Comment by mariajdab** [comment]
Perfect, I will add the suggestions.

**Review by pbiggar — APPROVED** [review]

**Comment by pbiggar** [comment]
Great, thank you! Nice job.

**Pull request merged**